ligo/src/token/fa2.mligo

```diff
@@ -0,0 +1,144 @@
+      }
+
+
+// -----------------------------------------------------------------
+// Transfer
+// -----------------------------------------------------------------
+
+let transfer_item (store, ti : storage * transfer_item): storage =
+  let transfer_one (acc, tx : storage * transfer_destination): storage =
+    let valid_from_ = check_sender (ti.from_, acc) in
+    if tx.amount = 0n then acc
+    else (
+      let valid_token_id =
+        if tx.token_id = unfrozen_token_id then tx.token_id
+        else if tx.token_id = frozen_token_id then (
+          if sender = store.admin then tx.token_id
+          else (failwith ("FROZEN_TOKEN_NOT_TRANSFERABLE"): token_id)
+        ) else (failwith ("FA2_TOKEN_UNDEFINED") : token_id)
+      in credit_to
+        ( tx.amount, tx.to_, valid_token_id
+        , debit_from (tx.amount, valid_from_, valid_token_id, acc)
+        )
+    )
+  in List.fold transfer_one ti.txs store
+
+let transfer (params, store : transfer_params * storage): return =
+  let store = ensure_not_migrated(store) in
+  let store = List.fold transfer_item params store in
+  (([] : operation list), store)
+
+
+// -----------------------------------------------------------------
+// Balance of
+// -----------------------------------------------------------------
+
+[@inline]
+let validate_token_type (token_id : token_id): token_id =
+  if (token_id = unfrozen_token_id || token_id = frozen_token_id) then
+    token_id
+  else (failwith ("FA2_TOKEN_UNDEFINED"): token_id)
+
+let balance_of (params, store : balance_request_params * storage): return =
+  let check_one (req : balance_request_item): balance_response_item =
+    let valid_token_id = validate_token_type(req.token_id) in
+    match Big_map.find_opt (req.owner, valid_token_id) store.ledger with
+      Some bal -> { request = req; balance = bal}
+    | None -> { request = req; balance = 0n}
+  in
+  let result = List.map check_one params.requests in
+  let transfer_operation = Tezos.transaction result 0mutez params.callback
+  in (([transfer_operation] : operation list), store)
+
+// -----------------------------------------------------------------
+// Token metadata registry entrypoint
+// -----------------------------------------------------------------
+
+let token_metadata_registry (params, store : token_metadata_registry_param * storage):return =
+  let store = ensure_not_migrated (store) in
+  let transfer_operation = Tezos.transaction store.token_address 0mutez params
+  in (([transfer_operation] : operation list), store)
+
+
+// -----------------------------------------------------------------
+// Update operators entrypoint
+// -----------------------------------------------------------------
+[@inline]
+let validate_operator_token (token_id : token_id): token_id =
+  if (token_id = unfrozen_token_id) then
+    token_id
+  else if (token_id = frozen_token_id) then
+    (failwith ("OPERATION_PROHIBITED"): token_id)
+  else
+    (failwith ("FA2_TOKEN_UNDEFINED"): token_id)
```

Comment:
Oh by the way, could you rename `assertEq0Or1` method in Lorentz version as well?

I think it would be nice to have as exact match between two versions of the contract as possible, and your name looks much better.

If we don't do this now, probably that name will remain there forever )

Reply:
Yeah, make sense. I've updated this now.